Use a decimal type for Transaction inflow, outflow, and amountMinor

jasonpetsod opened the pull request:
Install the decimal.js-light library to provide the decimal type. Convert all
call sites to pass in Decimal to Transaction().

Commit:
Change MutableTransactionRow to not store a Transaction in state

User inputs in a MutableTransactionRow don't necessarily map 1:1 with the
fields in a Transaction object; stop abusing Transaction objects as temporary
storage for user input. This will become especially true now that we're
starting to implement the actual data model: e.g. the 'account' and 'category'
fields will point to actual objects instead of being strings.

More immediately though, this change is required to start storing amountMinor
in Transaction instead of separate 'inflow' and 'outflow' fields.

Also stop passing 'id' to the ctor of Transaction when creating new
transactions. Population of the Transaction id is handled by addTransaction
since we use Firebase keys as the Transaction id.

Yes, this basically undoes commit 7efe39846b01c15cddb1ce98361cb1c0fe3dab63.

Issue: #15

Commit:
Replace inflow and outflow fields in Transaction with amountMinor

Add inflow() and outflow() methods that utilize amountMinor. Change components
to use these new methods.

Fix tests to not pass both inflow and outflow fields to Transaction().

Issue: #15

Commit:
Use a decimal type for Transaction inflow, outflow, and amountMinor

Install the decimal.js-light library to provide the decimal type. Convert all
call sites to pass in Decimal to Transaction().

Issue: #15

witriew commented on package.json:
```diff
@@ -9,6 +9,7 @@
     "babel-preset-es2015": "^6.22.0",
     "babel-preset-react": "^6.22.0",
     "babelify": "^7.3.0",
+    "decimal.js-light": "^2.2.0",
```
Is there a reason to not enforce precision to 2 (and other sorts of configs for Decimal)?

jasonpetsod replied:
I don't think we need to force a precision during intermediate computation. When we output to the browser we round to two decimal places anyway.

witriew replied:
Will we ever have a use-case where we need more precision during intermediate computation (I don't see us doing multiplication and division, only adds/subs)?  If not, I'd rather not have to ever worry about storing ridiculous floats.

jasonpetsod replied:
Filed issue #35.

witriew commented on src/Transaction.jsx:
```diff
@@ -1,36 +1,43 @@
+    } else if (inflow.isPositive()) {
+      this.amountMinor = inflow.times(100);
+    } else if (outflow.isPositive()) {
+      this.amountMinor = outflow.times(-100);
```
Use `.neg()`; http://mikemcl.github.io/decimal.js-light/#neg.

jasonpetsod replied:
Done.

witriew commented on src/Transaction.jsx:
```diff
@@ -1,36 +1,43 @@
-      // TODO: Stop using floats for money.
-      this.amountMinor = -1 * outflow * 100;
+    if (inflow.isZero() && outflow.isZero()) {
+      this.amountMinor = new Decimal('0');
```
Initialize this at the top and you can drop this test.

jasonpetsod replied:
Done.

witriew commented on src/Transaction.jsx:
```diff
@@ -1,36 +1,43 @@
-      // TODO: Stop using floats for money.
-      this.amountMinor = -1 * outflow * 100;
+    if (inflow.isZero() && outflow.isZero()) {
+      this.amountMinor = new Decimal('0');
```
Is there a reason to be doing the `int <-> Decimal` type conversion here?  I think this goes better when it gets written to/read from Firebase.

jasonpetsod replied:
Moved the conversion into MutableTransactionRow in 5119bbd1b62f4beaf3df5045c9174eb1a8255116.

Commit:
Move the logic to parse inflow and outflow to MutableTransactionRow

Change the Transaction ctor to only accept an amountMinor argument instead of
inflow and outflow. This makes MutableTransactionRow solely responsible for
parsing user input as it should be. Transaction now looks more like the data
model.

Commit:
Return Decimal from parseAmountMinor
